src/App.js

```diff
@@ -37,77 +128,429 @@ function App(props) {
+        // Use batch to make this operation atomic
+        const batch = writeBatch(db);
+
+        // Remove list from user lists array and replace with new list containing new list name
+        const newLists = userData.lists.map((list) => {
+            if (list === oldName) {
+                return newName;
+            } else {
+                return list
+            }
+        });
+        batch.update(doc(db, collectionName, userDocumentName), {
+            lists: newLists
+        });
+
+        // Delete all items under old list name
+        const oldItems = [...itemsData];
+        oldItems.forEach((item) => {
+            batch.delete(doc(db, collectionPath + subcollectionName, item.id));
+        })
+
+        // Add back all items under new list name
+        oldItems.forEach((item) => {
+            batch.set(doc(db, collectionPath + newName, item.id), {
+               id: item.id,
+               text: item.text,
+               completed: item.completed,
+               priority: item.priority,
+               created: item.created,
+            });
+        });
+
+        batch.commit();
+
+        // Update state to be on newly edited list
+        setSubcollectionName(newName);
     }
 
-    // Edit item
-    const onEditItem = (id, text, completed) => {
-        setData(data.map((item) => {
-            if (item.id === id) {
-                const newItem = {...item};
-                newItem.text = text;
-                newItem.completed = completed;
-                return newItem;
+    // Delete list
+    const onDeleteList = (name) => {
+        setUndoStack(undoStack.concat([{
+            type: UndoType.List,
+            op: UndoOp.Delete,
+            oldList: name,
+            newList: null,
+            data: itemsData,
+        }]));
+
+        // First delete all items in list
+        const batch = writeBatch(db);
+        itemsData.forEach((item) => {
+            batch.delete(doc(db, collectionPath + subcollectionName, item.id));
+        });
+        batch.commit();
+
+        // Then delete list itself from user array so it doesn't show in dropdown anymore
+        const newLists = userData.lists.filter(list => list !== subcollectionName);
+        updateDoc(doc(db, collectionName, userDocumentName), {
+            lists: newLists
+        })
+
+        // Update state since current subcollection list no longer exists
+        setSubcollectionName(userData.lists[0]);
+    }
+
+    // Undo operation so data is brought to previous state
+    const onUndo = () => {
+        const newStack = [...undoStack];
+        const undo = newStack.pop();
+
+        if (undo.type === UndoType.Item) { // Undo operation on item(s) within a list
+            const batch = writeBatch(db);
+
+            // Erase all current items from oldList
+            const q = query(collection(db, collectionPath + undo.oldList));
+            getDocs(q).then(res => {
+                res.docs.forEach((item) => {
```

Comment:
You may want to consider what the effect of such whole-sale undo will have on simultaneous users on other devices. Will they lose their intermediate changes between the operation and its undo?

Reply:
This is a good point. It is possible that user A could lose their changes when user B does an undo that takes the data back to user B's old data.